src/UserSpaceInstrumentation/RegisterState.cpp

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "UserSpaceInstrumentation/RegisterState.h"
+
+#include <cpuid.h>
+#include <elf.h>
+#include <errno.h>
+#include <sys/ptrace.h>
+#include <sys/uio.h>
+
+#include <string>
+
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/SafeStrerror.h"
+#include "absl/strings/str_format.h"
+
+namespace orbit_user_space_instrumentation {
+
+namespace {
+
+// Return the size of the XSave area on this cpu.
+[[nodiscard]] ErrorMessageOr<size_t> GetXSaveAreaSize() {
+  uint32_t eax = 0;
+  uint32_t ebx = 0;
+  uint32_t ecx = 0;
+  uint32_t edx = 0;
+  if (!__get_cpuid(1, &eax, &ebx, &ecx, &edx) || !(ecx & bit_XSAVE)) {
+    return ErrorMessage("XSAVE is not supported by Cpu.");
+  }
+  if (!__get_cpuid_count(0x0d, 0x00, &eax, &ebx, &ecx, &edx)) {
+    return ErrorMessage("XSAVE is not supported by Cpu.");
+  }
+  return static_cast<size_t>(ecx);
+}
+
+// Return offset of the YMMx registers inside the extended section of the XSave area.
+[[nodiscard]] ErrorMessageOr<size_t> GetAvxOffset() {
+  uint32_t eax = 0;
+  uint32_t ebx = 0;
+  uint32_t ecx = 0;
+  uint32_t edx = 0;
+  if (!__get_cpuid(0x01, &eax, &ebx, &ecx, &edx) || !(ecx & bit_AVX)) {
+    return ErrorMessage("AVX is not supported by Cpu.");
+  }
+  if (!__get_cpuid_count(0x0d, 0x02, &eax, &ebx, &ecx, &edx)) {
+    return ErrorMessage("AVX offset query failed.");
+  }
+  return static_cast<size_t>(ebx);
+}
+
+}  // namespace
+
+bool RegisterState::Hasx87DataStored() {
+  return (static_cast<uint64_t>(GetXSaveHeader()->xstate_bv) &
+          static_cast<uint64_t>(XSaveHeader::StateComponents::kX87)) != 0;
+}
+
+bool RegisterState::HasSseDataStored() {
+  return (static_cast<uint64_t>(GetXSaveHeader()->xstate_bv) &
+          static_cast<uint64_t>(XSaveHeader::StateComponents::kSse)) != 0;
+}
+
+bool RegisterState::HasAvxDataStored() {
+  return (static_cast<uint64_t>(GetXSaveHeader()->xstate_bv) &
+          static_cast<uint64_t>(XSaveHeader::StateComponents::kAvx)) != 0;
+}
+
+[[nodiscard]] ErrorMessageOr<void> RegisterState::BackupRegisters(pid_t tid) {
```

Comment:
One thing I find slightly weird of the interface of this class is that the `tid` is passed to `BackupResisters` but not to `RestoreRegisters`, instead the `tid_` stored by `BackupRegisters` is the used in `RestoreRegisters`. I can see the point but I would find it more consistent if both methods accepted the `tid` instead of storing it in the class. After all, `tid_` is not conceptually part of the "register state".

Reply:
Yes, I thought about that. I could not come up with any case in which injecting the registers of one thread into another leads to anything but desaster. So I chose to forbid it.